File: golang/599.minimum-index-sum-of-two-lists.go
func findRestaurant(list1 []string, list2 []string) []string {
	m, n := len(list1), len(list2)
	hash1 := make(map[string]int, m)

	for i := range list1 {
		hash1[list1[i]] = i
	}

	res := []string{}
	minVal := m + n

	for i := range list2 {
		if v, ok := hash1[list2[i]]; ok {
			if v+i < minVal {
				minVal = v + i
				res = []string{list2[i]}
			} else if v+i == minVal {
				res = append(res, list2[i])
			}
		}
	}

	return res
}

File: python/599.minimum-index-sum-of-two-lists.py
class Solution:
    def findRestaurant(self, list1: list[str], list2: list[str]) -> list[str]:

        hash1 = {k: v for v, k in enumerate(list1)}
        hash2 = {k: v for v, k in enumerate(list2)}

        minIJ = int(1e9)

        for s1 in list1:
            if s1 in hash2:
                minIJ = min(minIJ, hash1[s1] + hash2[s1])

        for s2 in list2:
            if s2 in hash1:
                minIJ = min(minIJ, hash1[s2] + hash2[s2])

        res = []
        for i in range(min(len(list1), minIJ + 1)):
            j = minIJ - i

            if j < len(list2) and list1[i] == list2[j]:
                res.append(list1[i])

        return res

